### PR TITLE
Attendance implementation

### PR DESCRIPTION
**Pull request by EvelynVoce**

### CODE REVIEW

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿using Dapper;
+using DevSoc.Eventathon.Attendance;
+using DevSoc.Eventathon.Calendars.Models;
+
+namespace DevSoc.Eventathon.Data;
+
+public class AttendanceService : IAttendanceService
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/AttendanceService.cs:6]: This is the data layer, so it's the repository pattern - rename to `AttendanceRepository` / `IAttendanceRepository`

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        using var connection = await _connectionManager.Open();
+        using var transaction = connection.BeginTransaction();
+
+        const string insertQuery = @"
+        INSERT INTO public.attendance
+        (""user_id"", ""event_id"")
+        VALUES(@UserId, @EventId);";
+
+        await connection.ExecuteAsync(insertQuery, attendanceData);
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/AttendanceService.cs]: we should be careful doing this, as we are reusing our HTTP layer class `AttendanceDefinition` as our data layer insertion class

File: src/DevSoc.Eventathon.Data/Migrations/Script0002-Attendance.sql [hunk 3]
```diff
@@ -0,0 +1,5 @@
+﻿CREATE TABLE Attendance (
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Data/Migrations/Script0002-Attendance.sql:1]: This script needs to be an `EmbeddedResource` and should copy to output if newer
(Right click > Properties to set this in Rider)

Comment (LukeBillo) [reply src/DevSoc.Eventathon.Data/Migrations/Script0002-Attendance.sql:1]: FYI this is why it doesn't migrate automatically

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs (hunk 1 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceService.cs:6]: I can't do that as dependency injection requires them to be in the IAttendanceService format

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs (hunk 2 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceService.cs]: What do you think would be better? Just splitting it out so it's {attendanceData.userID, attendanceData.eventID}?

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs (hunk 1 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceService.cs:6]: Nevermind, I think I was just doing something wrong

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs (hunk 2 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceService.cs]: I think I've sorted this. I've implemented it the same way we've implemented UsersRepository and UsersServices

File: src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+public class AttendanceRepository : IAttendanceRepository
+{
+    private readonly IConnectionManager _connectionManager;
+
+    public AttendanceRepository(IConnectionManager connectionManager)
+    {
+        _connectionManager = connectionManager;
+    }
+    
+    public async void RegisterAttendance(AttendanceData attendanceData)
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs]: Don't return `void` from an `async` function - you _must_ return the `Task` unless you are not `await`ing it.
You will end up `await`ing an empty value which will throw an exception.

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+﻿using DevSoc.Eventathon.Attendance.Models;
+using DevSoc.Eventathon.Calendars.Models;
+
+namespace DevSoc.Eventathon.Attendance;
+
+public class AttendanceService : IAttendanceService
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/AttendanceService.cs:6]: I'm on the fence about whether this service is really needed - you could just do the mapping in the controller and have a direct dependency on the `IAttendanceRepository`, but I think this is okay.
Up to you whether you keep the service or not.

File: src/DevSoc.Eventathon.Attendance/IAttendanceRepository.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using DevSoc.Eventathon.Attendance.Models;
+
+namespace DevSoc.Eventathon.Attendance;
+using DevSoc.Eventathon.Calendars.Models;
+
+public interface IAttendanceRepository
+{
+    void RegisterAttendance(AttendanceData definition);
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/IAttendanceRepository.cs]: This should return a `Task`

File: src/DevSoc.Eventathon.Attendance/AttendanceService.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    public void RegisterAttendance(AttendanceDefinition definition)
+    {
+        var attendanceData = new AttendanceData
+        {
+            UserId = definition.UserId,
+            EventId = definition.EventId
+        };
+
+        _attendanceRepository.RegisterAttendance(attendanceData);
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon.Attendance/AttendanceService.cs]: Need to `await` here, so should be an `async` method and return a `Task` here too.

File: src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs (hunk 4 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs]: What type of task do I return if it really could just be void though? Like in other functions we have Task< string > for example and I don't think I can just do Task< void >

Comment (LukeBillo) [reply src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs]: There's a non-generic type `Task` - `Task<void>` isn't valid

Comment (EvelynVoce) [reply src/DevSoc.Eventathon.Attendance/AttendanceRepository.cs]: How do you return that though? No matter what I do it keeps saying 'return type is void' even when I change it to Task and try to return something. If I change it to Task with no return statement it doesn't complain but then makes me question why I need change it to type Task in the first place

File: src/DevSoc.Eventathon/Controllers/AttendanceController.cs [hunk 8]
```diff
@@ -1,16 +1,30 @@
...
+    public AttendanceController(IAttendanceService attendanceService)
+    {
+        _attendanceService = attendanceService;
+    }
+    
     [HttpPost("api/attendance")]
     public Task<IActionResult> ConfirmAttendance([FromBody] AttendanceDefinition definition)
     {
-        // todo: implement attendance
+        _attendanceService.RegisterAttendance(definition);
```
Comment (LukeBillo) [inline src/DevSoc.Eventathon/Controllers/AttendanceController.cs:20]: `await _attendanceService.RegisterAttendace(definition);`?

**Review by LukeBillo — APPROVED** [review]

File: src/DevSoc.Eventathon/Controllers/AttendanceController.cs (hunk 8 above)
Comment (EvelynVoce) [reply src/DevSoc.Eventathon/Controllers/AttendanceController.cs:20]: I realise I forgot to change this in this PR but I have addressed this now

**Pull request merged**